### DISCUSSION TITLE
[front] 채팅 관련 API 요청, 응답 정리

### DISCUSSION THREAD

**Discussion by parktest0325** [3DPong/transcendence#46] [General]
#30 : 참여한, 전체 채팅 리스트 API 데이터 참고. 좀더 상세히 작성
▶️ : Request
◀️ : Response
⚠️ : 에러코드 반환 가능성 
📢 : 협의 필요
👍 : 협의안 투표
🌊 : option

필요한 API 경로, 요청 및 응답 데이터
이름이나 경로명은 자유롭게 변경해서 보내주셔도 됩니다. 

# 참여한 채팅방 리스트
###  채팅 리스트
  * <img height="200" alt="image" src="https://user-images.githubusercontent.com/52898964/226454724-3fdd913a-9bbc-4d08-b112-f0baaca546a1.png">
  * 요청한 사용자에 해당하는 채팅방리스트. 토큰이나 세션등으로 요청 유저 확인
  * ▶️ **Request** : GET /channels
  * ◀️ **Response** : Channel[] 형태로 JSON 응답
    * ⚠️ 요청한 사용자가 유효하지 않은경우 에러 (인증실패)
    * 🌊 채팅방 정렬 => 마지막 메시지? 아니면 type별?
```ts
interface Channel {
  id : number;
  name : string; 
  type : ChatType;          // type ChatType = "protected" | "private" | "public" | "dm";
  user : User
};

interface User {
  id: number;
  name: string;
  profileURL: string;
};
```

### 이미 참여한 채팅방 클릭 시 (1) - 유저 리스트
  * <img height="200" alt="image" src="https://user-images.githubusercontent.com/52898964/226455427-7bd469d7-7add-48f0-bd4a-b2a46c81b3f9.png">
  * 채팅 내용이 보여질 상세 채팅 페이지. 
  * 소프트딜리트된 사람까지 모두 전달해주면 클라이언트에서 필터링 후 userlist에 표시. 
     => 메시지로그에 아이디만 포함되어있는데, 대화중 나가면 프로필이미지랑 닉네임을 가져올 방법이 없기 때문에 필요함
  * ▶️ **Request** : GET /channels/[:channel_id]/users
  * ◀️ **Response** :  User[] 
    * ⚠️ 참여하지 않은 채팅방을 요청한 경우 에러 리턴, ⚠️ 유효하지 않은 요청유저
```ts
interface User {
  id: number;
  name: string;
  profileURL: string;
  role: UserRole;               // type UserRole = "owner" | "admin" | "user"; 
  // status?: UserStatus;      // type UserStatus = "online" | "offline" | "ingame" | "none";      // 🌊 보류
  is_deleted: boolean;       // 소프트딜리트 여부 (나간유저인 경우)
};
```

### 이미 참여한 채팅방 클릭 시 (2) - 메시지 리스트
  * 채팅 내용 리스트. 페이징 처리 
  * ▶️ **Request** : GET /channels/[:channel_id]/messages
  * ◀️ **Response** : Message[] 
    * ⚠️ 참여하지 않은 채팅방을 요청한 경우 에러 리턴, ⚠️ 유효하지 않은 요청유저
  * Date: 날짜형식   ex) 2023-03-20 20:15:37 등 자체적인 형식이 있음  
```ts
interface Message {
  id : number;                // MessageId
  senderId : number;     // Sender UserId 
  content: string;           // 메시지 내용
  createdAt: Date;          // 메시지 생성 시각. 타입을 어떻게 넣을지 협의 필요.
};
```

#    
#    
#    

# 전체 채팅방 리스트
### 채팅 리스트
  * 전체 채팅방 리스트 검색
      => 요청 유저를 확인할 필요가 없어서 좋은듯 👍 
      => 전체 채팅방에서 참여한 채팅방을 filter로 모두 비교해야함  ex) arr1.filter(x => arr2.includes(x));
  * ▶️ **Request** : GET /channels/add/[:string] 
    * [:string] 이 포함된 채팅방 배열 리턴
  * ◀️ **Response** : Channel[] 형태로 JSON 응답
    * ⚠️ 유저 인증이 필요없기 떄문에 서버에러를 제외하고는 에러가 발생하지 않을것으로 예상됨
  * Response 형태는 현재 채팅방 채팅리스트와 동일

### 클릭 시 채팅방 참여 
  * ▶️ **Request** : POST /channels/[:channelId]/join 
    * ` { "password" : string | null } `
    * `비밀번호가 없는 채팅방 진입의 경우`
      1.  "password" : null  👍  협의완료
  * ◀️ **Response** : 성공 실패 여부만 확인할 수 있으면 됨. (글로벌 채팅방 리스트를 가져올때 로컬 추가에 필요한 데이터를 전부 가지고있음)
    * ⚠️ 유효하지 않은 요청유저, ⚠️ 없는 채팅방, ⚠️ 비밀번호 틀림, ⚠️ 밴당한 페이지

# 채팅방 생성
### 채팅방 생성
  * <img  height="200" alt="image" src="https://user-images.githubusercontent.com/52898964/226468883-28883afe-e29d-4b13-b14e-019e5e6900cc.png">
  * ▶️ **Request** : PUT /channels/create 
    * ` {"title": string, "type": ChatType, "password": string | null, "inviteList" : number[] | null }`
    * 📢 요청 URL 적극변경필요..? PUT으로 하는게 맞나요?
  * ◀️ **Response** : Channel 타입으로 리턴. (배열이 아니라 Channel 객체 하나 이걸 받아서 바로 localChatList에 집어넣음)
    * ⚠️ 유효하지 않은 요청유저, ⚠️ 유효하지 않은 데이터(null값 입력)

### 초대할 유저 검색
  * 이건 친구추가 시 global 유저 검색 api를 사용하면 될듯 (유저리스트에서 id, nickname만 필요함) 

#    
#    
#    

# 채팅방 상세페이지 (without 웹소켓)
* 추가 : 상세 채팅방 페이지 진입시 관리자인 경우 뮤트리스트와 밴리스트도 함께 요청
### 뮤트 리스트 API, 밴 리스트 API
  * 해당 채팅방에서 뮤트된 유저 리스트
  * ▶️ **MuteList - Request** : GET /channels/[:channel_id]/mutelist 
  * ▶️ **BanList - Request** : GET /channels/[:channel_id]/banlist 
  * ◀️ **Response** : FlagUser[] 형태로 JSON 응답
    * ⚠️ 해당 채팅방의 관리자인지 체크 필요함
```ts
interface FlagUser {
  end_at: Date;
  user : User == {
    id: number;
    name: string;
    profileURL: string;
  }
};
```

### 채팅방 유저 리스트,  채팅방 메시지 (페이징처리)
* 얘는 처음 접속할때 이미 위에서 정의함 

### 유저관리 (킥, 밴, 뮤트, 언밴, 언뮤트)
킥 밴 뮤트 할때 시간도 전달

**Comment by parktest0325** [comment]
@kyu-baek 규림님 채팅방 리스트와 생성때 필요한 API 다시 정리해봤습니다 혹시 다른의견이 있으시거나 📢 부분에서 의견이 있으시면 코멘트로 달아주시면 될것같습니다.

**Reply by parktest0325** [reply]
일단 상세페이지 제외하고 총 5개의 API 입니다.

**Comment by parktest0325** [comment]
최우선 협의사항은 createdAt이 어떤 형태로 DB에 저장되고, Response 될때 어떤 값이 전달되는지 알아야 할것같습니다. 만약 서버에서 DB에서 긁어와서 리턴할때 unix 타임스탬프(정수) 값이라면 서버측에서 unix 타임스탬프 값을 변경하는것보다 클라이언트에서 변환하는게 좋을것같아보여요

**Comment by kyu-baek** [comment]
확인했습니다!

**Comment by parktest0325** [comment]
@kyu-baek 밴리스트, 뮤트리스트 API 추가했습니다

**Comment by kyu-baek** [comment]
### 참여한 채팅방 리스트
▶️ Request : GET /channels/[:channel_id]
◀️ Response : { users: User[], messages: Messages[] } 형태로 JSON 응답

의 경우 pagination 으로 처리하면 

<img width="841" alt="image" src="https://user-images.githubusercontent.com/69143394/226921843-d7246171-7c87-40b5-8f31-2deb0a5f01eb.png">

예를 들어 이런 식으로 take(요청시 보내는 게시물 수), skip(페이징을 위해 스킵하는 수) 쿼리를 보내주게 됩니다.
하지만 그경우 여러번 커서가 위로 올라가거나 할 때마다 재요청이 될텐데 그 때마다 user객체들이 불필요하게 재검색 됩니다. 

1. 백에서 한번에 모든 message log 데이터와 참여한 user 객체들을 넘겨준다.

2. users api 요청과 message log api 요청을 분리하고 pagination 을 사용한다.
   a. 처음 방에 들어갈 때는 users, message 이렇게 총 2개의 api 요청. 이후부터는 message api 만 요청
 
둘다 어렵지는 않을 것 같습니다. 동현님께서 선택해주시면 제가 맞춰서 수정하겠습니다.

**Comment by kyu-baek** [comment]
### 클릭 시 채팅방 참여

body 에 
channel_id , password (null 가능)
dto 를 따로 선언해서 받아올 예정이라.
이렇게 두개 고정으로 받는 걸로 해주시면 감사합니다.

**Comment by parktest0325** [comment]
userapi와 messagelog api가 나눠진 2번으로 결정하고, take와 skip을 제가 전달해드리는 방식으로 하겠습니다.
일단 연결할땐 take=20, skip=0으로 보낼예정이고 스크롤 처리하면 가져온 데이터 값 + 새로온 메시지 수 까지 계산해서 take=20, skip=계산결과값 으로 요청드릴게요 
채팅방 참여 시 channel_id, password(널 가능) 값으로 해서 디스커션 수정하겠습니다

**Comment by kyu-baek** [comment]
이미 참여한 채팅방 클릭 시 (1) - 유저 리스트
```
[
    {
        "role": "user",
        "deleted_at": null,
        "user": {
            "user_id": 15,
            "nickname": "correspondent",
            "profile_url": "https://loremflickr.com/640/480/people"
        }
    },
    {
        "role": "user",
        "deleted_at": "2023-03-27T04:07:30.021Z",
        "user": {
            "user_id": 51,
            "nickname": "cargo",
            "profile_url": "https://loremflickr.com/640/480/people"
        }
    },
    {
        "role": "owner",
        "deleted_at": null,
        "user": {
            "user_id": 82,
            "nickname": "sonata",
            "profile_url": "https://loremflickr.com/640/480/people"
        }
    },
    {
        "role": "user",
        "deleted_at": null,
        "user": {
            "user_id": 99,
            "nickname": "innovation",
            "profile_url": "https://loremflickr.com/640/480/people"
        }
    }
]
```

**Comment by kyu-baek** [comment]
https://gusrb3164.github.io/web/2021/09/25/websocket-data/
